### PR TITLE
do not sync persistent store when get a fake position

### LINKED ISSUE #396: 执行大事务时出现binlog解析失败
现象：Err: table id 267724937: invalid table id, no corresponding table map event"
```
[2019/06/25 18:04:19] [error] binlogsyncer.go:639 connection was bad
[2019/06/25 18:04:20] [info] binlogsyncer.go:581 begin to re-sync from (mysql-bin.003014, 923556481)
[2019/06/25 18:04:20] [info] binlogsyncer.go:211 register slave for master server 
[2019/06/25 18:04:20] [error] binlogsyncer.go:239 kill connection 2818644069 error ERROR 1094 (HY000): Unknown thread id: 2818644069
[2019/06/25 18:04:20] [info] binlogsyncer.go:245 kill last connection id 2818644069
[2019/06/25 18:04:20] [info] binlogsyncer.go:731 rotate to (mysql-bin.003014, 923556481)
[2019/06/25 18:15:41] [info] sync.go:66 rotate binlog to (mysql-bin.003014, 923556481)
```
当binlog同步过程中出现网络错误导致重连resync的时候，mysql会向canal发送rorate 事件更新当前的Pos. 如果此时重启了canal。canal将从当前保存的这个位置开始继续同步。但是由于当前这条binlog所对应的table id获取不到，会出现no corresponding table map even 错误。
对应的binlog如下
```
| mysql-bin.003014 | 906327751 | Table_map      | 341013306 |   906327860 | table_id: 267724937 (bilibili_upcrm.up_base_info)            |
| mysql-bin.003014 | 906327860 | Delete_rows_v1 | 341013306 |   906328894 | table_id: 267724937                                          |
| mysql-bin.003014 | 906328894 | Delete_rows_v1 | 341013306 |   906329854 | table_id: 267724937                                          |
| mysql-bin.003014 | 906329854 | Delete_rows_v1 | 341013306 |   906330834 | table_id: 267724937                                          |
| mysql-bin.003014 | 906330834 | Delete_rows_v1 | 341013306 |   906331808 | table_id: 267724937                                          |
| mysql-bin.003014 | 906331808 | Delete_rows_v1 | 341013306 |   906332738 | table_id: 267724937                                          |
| mysql-bin.003014 | 906332738 | Delete_rows_v1 | 341013306 |   906333714 | table_id: 267724937                                          |
| mysql-bin.003014 | 906333714 | Delete_rows_v1 | 341013306 |   906334692 | table_id: 267724937
... 
```
由于执行了delete from table 操作，导致产生大量binlog。table map event已经被解析完成，同时由于resync，导致pos被更新到 table map event事件之后。重启之后获取不到table map 就会导致 invalid table id 的报错

### PR DESCRIPTION
**Pull request by lintanghui**
fixed #396

### CODE REVIEW

**Comment by lintanghui** [comment]
@siddontang PTAL

**Comment by siddontang** [comment]
PTAL @GregoryIan

File: canal/sync.go [hunk 1]
```diff
@@ -65,6 +65,10 @@ func (c *Canal) runSyncBinlog() error {
 		// TODO: If we meet any DDL query, we must save too.
 		switch e := ev.Event.(type) {
 		case *replication.RotateEvent:
+			// if event pos equal to local pos.maybe be fake rotate event,do nothing.
+			if string(e.NextLogName) == pos.Name && uint32(e.Position) == pos.Pos {
```
Comment (csuzhangxc) [inline canal/sync.go]: in MySQL, a fake rotate event is an event with _timestamp_ equals to **0** (see <https://github.com/mysql/mysql-server/blob/8cc757da3d87bf4a1f07dcfb2d3c96fed3806870/sql/rpl_binlog_sender.cc#L899>), and also with _log pos_ equals to **0** in the implementation (see <https://github.com/mysql/mysql-server/blob/8cc757da3d87bf4a1f07dcfb2d3c96fed3806870/sql/rpl_binlog_sender.cc#L906>).

I don't know whether some corner cases exist if checking by `string(e.NextLogName) == pos.Name && uint32(e.Position) == pos.Pos`.

Comment (siddontang) [reply canal/sync.go]: Friendly ping @lintanghui

Comment (lintanghui) [reply canal/sync.go]: > in MySQL, a fake rotate event is an event with _timestamp_ equals to **0** (see https://github.com/mysql/mysql-server/blob/8cc757da3d87bf4a1f07dcfb2d3c96fed3806870/sql/rpl_binlog_sender.cc#L899), and also with _log pos_ equals to **0** in the implementation (see https://github.com/mysql/mysql-server/blob/8cc757da3d87bf4a1f07dcfb2d3c96fed3806870/sql/rpl_binlog_sender.cc#L906).
> 
> I don't know whether some corner cases exist if checking by `string(e.NextLogName) == pos.Name && uint32(e.Position) == pos.Pos`.

yes, timestamp 0 and pos 0 is set in event header, but in event body, pos and file name is also be set,   https://github.com/mysql/mysql-server/blob/8cc757da3d87bf4a1f07dcfb2d3c96fed3806870/sql/rpl_binlog_sender.cc#L909  . in sync.go we just get event body and compare it.  we can also do such below
`ev.Header.Pos==0 {continue }`

Comment (csuzhangxc) [reply canal/sync.go]: @lintanghui I know, in this special case `string(e.NextLogName) == pos.Name && uint32(e.Position) == pos.Pos` is right, but I don't know whether some **corner cases** exist.

BTW, will you update it to `ev.Header.Pos==0 {continue }`?

Comment (lintanghui) [reply canal/sync.go]: fine, it seem better to compare header pos.

**Comment by lintanghui** [comment]
PTAL @GregoryIan  . it's that ok to be merge?

**Comment by mefcorvi** [comment]
I've faced with an issue where binlog filename is wrong after couple of rotate events.
```
[info] sync.go:25 start sync binlog at binlog file (mysql-bin.000006, 632)
[info] binlogsyncer.go:731 rotate to (mysql-bin.000006, 632)
[info] binlogsyncer.go:731 rotate to (mysql-bin.000007, 4)
[info] binlog-position.go:137 binlog position updated to ((mysql-bin.000006, 1044), <nil>)
```
The last line in the log is written by my code in the `OnXID` handler.

It seems that these two rotate events from the log were fake and were ignored in `sync.go` so filename hasn't been updated in the `pos` variable. Are you sure that fake rotate events must be ignored completely?

**Pull request merged**